File: scripts/solar/solar-common/actor/state/SolarActorAttributeState.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 1,["9"] = 1,["10"] = 2,["11"] = 2,["12"] = 3,["13"] = 3,["14"] = 4,["15"] = 4,["16"] = 5,["17"] = 5,["18"] = 6,["19"] = 6,["20"] = 7,["21"] = 7,["22"] = 8,["23"] = 8,["24"] = 9,["25"] = 9,["26"] = 13,["27"] = 13,["28"] = 13,["30"] = 29,["31"] = 30,["35"] = 34,["36"] = 35,["37"] = 36,["38"] = 37,["39"] = 39,["40"] = 40,["43"] = 43,["44"] = 44,["45"] = 46,["46"] = 47,["47"] = 46,["48"] = 37,["49"] = 52,["50"] = 52,["51"] = 52,["52"] = 54,["53"] = 55,["54"] = 54,["55"] = 52,["56"] = 52,["57"] = 59,["58"] = 60,["59"] = 61,["61"] = 59,["62"] = 64,["63"] = 65,["64"] = 66,["65"] = 66,["66"] = 66,["67"] = 67,["68"] = 66,["69"] = 66,["70"] = 64,["71"] = 70,["72"] = 72,["73"] = 73,["74"] = 73,["75"] = 73,["76"] = 74,["77"] = 73,["78"] = 73,["79"] = 70,["80"] = 28,["81"] = 80,["82"] = 82,["85"] = 85,["86"] = 86,["87"] = 86,["88"] = 87,["89"] = 88,["93"] = 92,["95"] = 97,["96"] = 98,["97"] = 99,["99"] = 102,["100"] = 103,["101"] = 105,["102"] = 107,["103"] = 80,["104"] = 115,["105"] = 116,["106"] = 117,["107"] = 118,["108"] = 120,["110"] = 120,["112"] = 120,["113"] = 121,["114"] = 122,["116"] = 125,["117"] = 126,["118"] = 127,["119"] = 128,["121"] = 130,["122"] = 131,["124"] = 131,["126"] = 131,["127"] = 132,["128"] = 133,["132"] = 138,["133"] = 139,["134"] = 140,["135"] = 141,["136"] = 142,["137"] = 143,["138"] = 144,["140"] = 146,["141"] = 147,["142"] = 148,["143"] = 149,["146"] = 152,["151"] = 157,["152"] = 158,["154"] = 159,["155"] = 159,["156"] = 160,["157"] = 161,["158"] = 162,["159"] = 162,["160"] = 163,["161"] = 164,["162"] = 165,["164"] = 167,["165"] = 168,["166"] = 169,["167"] = 170,["168"] = 171,["171"] = 174,["174"] = 159,["178"] = 179,["179"] = 115,["180"] = 22,["181"] = 23,["182"] = 25});
local ____exports = {}
local ____trigger = require("solar.solar-common.w3ts.handles.trigger")
local Trigger = ____trigger.Trigger
local ____Actor = require("solar.solar-common.actor.Actor")
local Actor = ____Actor.default
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
local ____ActorBuffUtil = require("solar.solar-common.actor.util.ActorBuffUtil")
local ActorBuffUtil = ____ActorBuffUtil.default
local ____UnitStateUtil = require("solar.solar-common.util.unit.UnitStateUtil")
local UnitStateUtil = ____UnitStateUtil.default
local ____AttributeUtil = require("solar.solar-common.util.system.AttributeUtil")
local AttributeUtil = ____AttributeUtil.default
local ____UnitAttributeState = require("solar.solar-common.attribute.UnitAttributeState")
local UnitAttributeState = ____UnitAttributeState.default
local ____SingletonUtil = require("solar.solar-common.util.lang.SingletonUtil")
local SingletonUtil = ____SingletonUtil.default
local ____BaseUtil = require("solar.solar-common.util.BaseUtil")
local BaseUtil = ____BaseUtil.default
____exports.default = __TS__Class()
local SolarActorAttributeState = ____exports.default
SolarActorAttributeState.name = "SolarActorAttributeState"
function SolarActorAttributeState.prototype.____constructor(self)
    if SingletonUtil:notFirstTime(____exports.default) then
        print("不能重复new SolarActorAttributeState()")
        return
    end
    --- buff 属性更新到单位属性
    local trigger2 = __TS__New(Trigger)
    local noUpdateAttributeTime = 0
    trigger2:registerTimerEvent(0.99, true)
    trigger2:addAction(function()
        if Actor._sl_needUpdateAttribute == false and noUpdateAttributeTime < 5 then
            noUpdateAttributeTime = noUpdateAttributeTime + 1
            return
        end
        Actor._sl_needUpdateAttribute = false
        noUpdateAttributeTime = 0
        DataBase:forUnitSolarDatas(function(____, u, solarData)
            ____exports.default:refreshActorAttributes2UnitSolarAttribute(u)
        end)
    end)
    se:on(
        "属性刷新",
        function()
            DataBase:forUnitSolarDatas(function(____, u, solarData)
                ____exports.default:refreshActorAttributes2UnitSolarAttribute(u)
            end)
        end
    )
    ActorBuffUtil:addAnyActorBuffCreatedListener(function(____, buff)
        if buff.attribute then
            ____exports.default:refreshActorAttributes2UnitSolarAttribute(buff.unit)
        end
    end)
    se:onUnitPickupItem(function(e)
        local trigUnit = e.trigUnit
        BaseUtil.runLater(
            0.1,
            function()
                ____exports.default:refreshActorAttributes2UnitSolarAttribute(trigUnit)
            end
        )
    end)
    se:onUnitDropItem(function(e)
        local trigUnit = e.trigUnit
        BaseUtil.runLater(
            0.1,
            function()
                ____exports.default:refreshActorAttributes2UnitSolarAttribute(trigUnit)
            end
        )
    end)
end
function SolarActorAttributeState.refreshActorAttributes2UnitSolarAttribute(self, unitHandle)
    if not UnitStateUtil:isAlive(unitHandle) then
        return
    end
    local attributes = ____exports.default:getUnitAllActorAttributes(unitHandle)
    local ____opt_0 = DataBase:getUnitSolarData(unitHandle, false)
    local oldAttr = ____opt_0 and ____opt_0._SL_totalActorsSolarAttribute
    if oldAttr == nil then
        if attributes == nil or #attributes == 0 then
            return
        end
    end
    local totalAttribute = AttributeUtil:sumAttributes(attributes)
    --- 属性 系统
    local solarData = DataBase:getUnitSolarData(unitHandle)
    if not solarData._SL_solarAttribute then
        solarData._SL_solarAttribute = {}
    end
    AttributeUtil:subtract(solarData._SL_solarAttribute, oldAttr)
    AttributeUtil:add(solarData._SL_solarAttribute, totalAttribute)
    solarData._SL_totalActorsSolarAttribute = totalAttribute
    UnitAttributeState:refreshUnitSolarAttribute(unitHandle)
end
function SolarActorAttributeState.getUnitAllActorAttributes(self, unit)
    local attributeArray = nil
    local solarData = DataBase:getUnitSolarData(unit, false)
    if solarData ~= nil then
        local ____opt_2 = solarData._SL_solarActorUnit
        if ____opt_2 ~= nil then
            ____opt_2 = ____opt_2.attribute
        end
        local attribute = ____opt_2
        if attribute ~= nil then
            attributeArray = {attribute}
        end
        local actorAbilitys = solarData._SL_solarActorAbilitys
        if actorAbilitys ~= nil then
            if attributeArray == nil then
                attributeArray = {}
            end
            for abilityTemplateKey in pairs(actorAbilitys) do
                local ____opt_4 = actorAbilitys[abilityTemplateKey]
                if ____opt_4 ~= nil then
                    ____opt_4 = ____opt_4.attribute
                end
                local attribute = ____opt_4
                if attribute ~= nil then
                    attributeArray[#attributeArray + 1] = attribute
                end
            end
        end
        local _SL_solarActorBuffSet = solarData._SL_solarActorBuffs
        if _SL_solarActorBuffSet then
            for ____, actorBuff in ipairs(_SL_solarActorBuffSet) do
                local resultAttribute = actorBuff.attribute
                if resultAttribute ~= nil then
                    if attributeArray == nil then
                        attributeArray = {}
                    end
                    if #____exports.default.buffAttributeHandlers > 0 then
                        resultAttribute = __TS__ObjectAssign({}, actorBuff.attribute)
                        for ____, buffAttributeHandler in ipairs(____exports.default.buffAttributeHandlers) do
                            resultAttribute = buffAttributeHandler(nil, actorBuff, resultAttribute)
                        end
                    end
                    attributeArray[#attributeArray + 1] = resultAttribute
                end
            end
        end
    end
    if ____exports.default.enableItemAttribute then
        local invSize = UnitInventorySize(unit)
        do
            local i = 0
            while i < invSize do
                local item = UnitItemInSlot(unit, i)
                if IsHandle(item) then
                    local ____opt_6 = DataBase:getItemSolarData(item, false)
                    local actorItem = ____opt_6 and ____opt_6._SL_solarActorItem
                    if actorItem and actorItem.attribute ~= nil then
                        if attributeArray == nil then
                            attributeArray = {}
                        end
                        local resultAttribute = actorItem.attribute
                        if #____exports.default.itemAttributeHandlers > 0 then
                            resultAttribute = __TS__ObjectAssign({}, actorItem.attribute)
                            for ____, itemAttributeHandler in ipairs(____exports.default.itemAttributeHandlers) do
                                resultAttribute = itemAttributeHandler(nil, actorItem, resultAttribute)
                            end
                        end
                        attributeArray[#attributeArray + 1] = resultAttribute
                    end
                end
                i = i + 1
            end
        end
    end
    return attributeArray
end
SolarActorAttributeState.itemAttributeHandlers = {}
SolarActorAttributeState.buffAttributeHandlers = {}
SolarActorAttributeState.enableItemAttribute = true
return ____exports

File: src/solar/solar-common/actor/state/SolarActorAttributeState.ts
import {Trigger} from "@/trigger";
import Actor from "@/Actor";
import DataBase from "@/DataBase";
import ActorBuffUtil from "@/ActorBuffUtil";
import UnitStateUtil from "@/UnitStateUtil";
import AttributeUtil from "@/AttributeUtil";
import UnitAttributeState from "@/UnitAttributeState";
import SingletonUtil from "@/SingletonUtil";
import BaseUtil from "@/BaseUtil";
import ActorItem from "@/ActorItem";
import ActorBuff from "@/ActorBuff";

export default class SolarActorAttributeState {

    /**
     * 演员物品属性处理器 (可以在这里做装备加成等增幅)
     *
     * resultAttribute 返回的最终给单位加上的属性 可以与ActorItem.attribute 不一样 以免改到基础属性
     * resultAttribute 初始为复制的ActorItem.attribute
     * 处理器通常应直接return resultAttribute
     */
    static itemAttributeHandlers: ((actor: ActorItem, resultAttribute: AppAttribute) => AppAttribute)[] = [];
    static buffAttributeHandlers: ((actor: ActorBuff, resultAttribute: AppAttribute) => AppAttribute)[] = [];
    /** 启用装备属性 (部分Orpg模拟穿戴装备的可以使用false 禁用物品栏里的演员物品属性效果) */
    static enableItemAttribute = true;


    constructor() {
        if (SingletonUtil.notFirstTime(SolarActorAttributeState)) {
            print("不能重复new SolarActorAttributeState()")
            return;
        }
        /**buff 属性更新到单位属性*/
        let trigger2 = new Trigger();
        let noUpdateAttributeTime = 0;
        trigger2.registerTimerEvent(0.99, true)
        trigger2.addAction(() => {
            //每秒定时更新  超过5秒没更新时 强制更新 保底机制可以提高框架鲁棒性
            if (Actor._sl_needUpdateAttribute == false && noUpdateAttributeTime < 5) {
                noUpdateAttributeTime++;
                return
            }
            Actor._sl_needUpdateAttribute = false;
            noUpdateAttributeTime = 0;
            //遍历所有
            DataBase.forUnitSolarDatas((u, solarData: AppData) => {
                SolarActorAttributeState.refreshActorAttributes2UnitSolarAttribute(u)
            })

        });
        //实时刷新
        se.on("属性刷新", () => {
            //遍历所有
            DataBase.forUnitSolarDatas((u, solarData: AppData) => {
                SolarActorAttributeState.refreshActorAttributes2UnitSolarAttribute(u)
            })
        })
        //实时刷新
        ActorBuffUtil.addAnyActorBuffCreatedListener(buff => {
            if (buff.attribute) {
                SolarActorAttributeState.refreshActorAttributes2UnitSolarAttribute(buff.unit)
            }
        });
        se.onUnitPickupItem(e => {
            let trigUnit = e.trigUnit;
            BaseUtil.runLater(0.1, () => {
                SolarActorAttributeState.refreshActorAttributes2UnitSolarAttribute(trigUnit);
            });
        });
        se.onUnitDropItem(e => {
            //延迟 刷新 此时可能装备还在身上 或其他物品修改逻辑
            let trigUnit = e.trigUnit;
            BaseUtil.runLater(0.1, () => {
                SolarActorAttributeState.refreshActorAttributes2UnitSolarAttribute(trigUnit);
            });
        });

    }

    static refreshActorAttributes2UnitSolarAttribute(unitHandle: unit) {
        //死亡单位不刷新
        if (!UnitStateUtil.isAlive(unitHandle)) {
            return;
        }
        let attributes: AppAttribute[] = SolarActorAttributeState.getUnitAllActorAttributes(unitHandle);
        let oldAttr: AppAttribute = DataBase.getUnitSolarData(unitHandle, false)?._SL_totalActorsSolarAttribute;
        if (oldAttr == null) {//如果之前没有旧buff属性 则退出后续逻辑  否则还需将之前的旧属性 移除
            if (attributes == null || attributes.length == 0) {
                return
            }
        }
        let totalAttribute: SolarAttribute = AttributeUtil.sumAttributes(attributes);

        /**
         属性 系统
         */
        let solarData = DataBase.getUnitSolarData(unitHandle);
        if (!solarData._SL_solarAttribute) {
            solarData._SL_solarAttribute = {}
        }

        AttributeUtil.subtract(solarData._SL_solarAttribute, oldAttr);
        AttributeUtil.add(solarData._SL_solarAttribute, totalAttribute);

        solarData._SL_totalActorsSolarAttribute = totalAttribute;
        //发送属性改变事件
        UnitAttributeState.refreshUnitSolarAttribute(unitHandle)
    }


    /**
     * 获取单位身上所有演员 的属性值
     * @param unit
     */
    static getUnitAllActorAttributes(unit: unit): AppAttribute[] {
        let attributeArray: AppAttribute[] = null;
        let solarData = DataBase.getUnitSolarData(unit, false);
        if (solarData != null) {
            //单位
            let attribute: AppAttribute = solarData._SL_solarActorUnit?.attribute;
            if (attribute != null) {
                attributeArray = [attribute];
            }
            //技能
            let actorAbilitys = solarData._SL_solarActorAbilitys;
            if (actorAbilitys != null) {
                if (attributeArray == null) {
                    attributeArray = [];
                }
                for (let abilityTemplateKey in actorAbilitys) {
                    let attribute = actorAbilitys[abilityTemplateKey]?.attribute;
                    if (attribute != null) {
                        attributeArray.push(attribute)
                    }
                }
            }
            //演员buff
            let _SL_solarActorBuffSet: ActorBuff[] = solarData._SL_solarActorBuffs;
            if (_SL_solarActorBuffSet) {
                for (let actorBuff of _SL_solarActorBuffSet) {
                    let resultAttribute = actorBuff.attribute
                    if (resultAttribute != null) {
                        if (attributeArray == null) {
                            attributeArray = [];
                        }
                        if (SolarActorAttributeState.buffAttributeHandlers.length > 0) {
                            resultAttribute = {...actorBuff.attribute}
                            for (let buffAttributeHandler of SolarActorAttributeState.buffAttributeHandlers) {
                                resultAttribute = buffAttributeHandler(actorBuff, resultAttribute)
                            }
                        }
                        attributeArray.push(resultAttribute)
                    }
                }
            }
        }
        if (SolarActorAttributeState.enableItemAttribute) {
            let invSize = UnitInventorySize(unit)
            for (let i = 0; i < invSize; i++) {
                let item = UnitItemInSlot(unit, i);
                if (IsHandle(item)) {
                    let actorItem: ActorItem = DataBase.getItemSolarData(item, false)?._SL_solarActorItem;
                    if (actorItem && actorItem.attribute != null) {
                        if (attributeArray == null) {
                            attributeArray = [];
                        }
                        let resultAttribute = actorItem.attribute;
                        if (SolarActorAttributeState.itemAttributeHandlers.length > 0) {
                            resultAttribute = {...actorItem.attribute}
                            for (let itemAttributeHandler of SolarActorAttributeState.itemAttributeHandlers) {
                                resultAttribute = itemAttributeHandler(actorItem, resultAttribute)
                            }
                        }
                        attributeArray.push(resultAttribute);
                    }
                }
            }
        }
        return attributeArray;
    }
}